src/sqlfluff/dialects/dialect_tsql.py

```diff
@@ -256,12 +256,6 @@
         Ref("DatatypeSegment"),
         Sequence(Ref("EqualsSegment"), Ref("ExpressionSegment"), optional=True),
     ),
-    FunctionNameIdentifierSegment=RegexParser(
-        r"[A-Z][A-Z0-9_]*|\[[A-Z][A-Z0-9_]*\]",
-        CodeSegment,
-        name="function_name_identifier",
-        type="function_name_identifier",
-    ),
```

Comment:
Would ljst adding an exclude list here work instead, similar to how we define`NakedIdentifierSegment` and `DatatypeIdentifierSegment`?

```py
    FunctionNameIdentifierSegment=SegmentGenerator(
        # Generate the anti template from the set of reserved keywords
        # minus the function names that are reserved words.
        lambda dialect: RegexParser(
            r"[A-Z][A-Z0-9_]*|\[[A-Z][A-Z0-9_]*\]",
            CodeSegment,
            name="function_name_identifier",
            type="function_name_identifier",
            anti_template=r"^("
            + r"|".join(
                dialect.sets("reserved_keywords")
                - {
                    "COALESCE",
                    "CONVERT",
                    "CURRENT_TIMESTAMP",
                    "CURRENT_USER",
                    "LEFT",
                    "NULLIF",
                    "RIGHT",
                    "SESSION_USER",
                    "SYSTEM_USER",
                }
            )
            + r")$",
        )
    ),
```

This would allow us to keep the parse tree the same as other dialects rather than introducing another level of Identifier or Keyword. That would also result in a lot less differences.

Reply:
@tunetheweb that's the direction I was thinking once I hit the critical error, but I haven't worked with the anti_template pattern yet.  I'll try it out!

Reply:
Looks much better and cleaner now! Hopefully tests will all pass too (they should IMHO)